### DISCUSSION TITLE
Design: use `and_then`?

### DISCUSSION THREAD

**Discussion by infiniteregrets** [metalbear-co/mirrord#475] [Polls]
With and_then
```rs
let pod_label = deployment
            .spec
            .and_then(|spec| spec.template.metadata)
            .and_then(|metadata| metadata.labels)
            .and_then(|labels| labels.get("app").cloned())
            .ok_or_else(|| {
                LayerError::DeploymentNotFound(format!(
                    "Label for deployment: {}, not found!",
                    self.deployment.clone()
                ))
            })?;
 ```
  
  Without and_then
  ```rs
  let pod_label = || -> Option<String> {
            deployment                
            .spec?
            .template
            .metadata?
            .labels?
            .get("app")
            .cloned()
        }()
        .ok_or_else(|| {
            LayerError::DeploymentNotFound(format!(
                "Label for deployment: {}, not found!",
                self.deployment.clone()
            ))
        })?;
```
  
  and_then just helps doing everything in chains, no use of `?` or a closure. 
  
  we need to use closures in other places where multiple things return Option

**Comment by meowjesty** [comment]
Or maybe for this case (heavy use of `Option`), just `try` block as you've suggested.

**Comment by infiniteregrets** [comment]
do too many `and_thens` make code look ugly?

**Reply by eyalb181** [reply]
Is the `and_then`s' advantage in this example that it makes the closure unnecessary? Because the actual code we were discussing has the closure anyway, since it uses `?` inside one of the `and_then`.

**Reply by infiniteregrets** [reply]
oh yeah sorry ignore that, the closure is not needed at all

**Reply by eyalb181** [reply]
Wait so ignore the closure in the "without and_then" option?

**Reply by infiniteregrets** [reply]
nope the code we were looking at yesterday had and_thens inside the closure, we didn't need that extra closure there
